feat: Allow errors to be formatted via codespan

Marwes opened the pull request:
Codespan adds in highlighted source to error messages which gives more direct feedback to users as they do not need to trace out the line and column numbers themselves.

cc #4264 #1380

Marwes commented on libflux/flux-core/src/semantic/tests.rs:
```diff
@@ -3655,16 +3692,27 @@ error @3:17-3:18: undefined identifier y",
+              ┌─ :2:13
+              │
+            2 │             isType(v: {}, type: "record")
+              │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ {} is not Basic (argument v)
```
Due to how constraints are generated for call expressions we unfortunately do not get more precise information atm, should be fixable though.

wolffcm replied:
The formatting is really nice. I wonder about how errors are rendered in the UI though, is there a way to get the more concise message?

onelson replied:
If it's not possible to present a multi-line error through the UI, I'd bet we could branch the reporting based on a config option.

Marwes replied:
Yep, we can just branch on it and omit the source display. codespan has some support for alternate displays as well https://docs.rs/codespan-reporting/0.11.1/codespan_reporting/term/enum.DisplayStyle.html

Marwes commented:
![Screenshot from 2021-12-03 14-06-37](https://user-images.githubusercontent.com/957312/144607503-c6094930-d148-4594-9222-f32aadb9299a.png)

Rebased and updated the repl/cmd line tool to use the codespan format with coloring.

scbrickley commented:
This patch is not actually changing the way errors are reported, it's just introducing the capability to use the codespan formatting, right?

If that's the case, this looks fine to me. We can figure out how to (as @onelson said) branch this so that we don't break things in the UI.

nathanielc commented on libflux/include/influxdata/flux.h:
```diff
@@ -104,7 +107,7 @@ void flux_free_stateful_analyzer(struct flux_stateful_analyzer_t *);
 // flux_analyze_with will analyze the ast snippet using the flux_stateful_analyzer_t and produce
 // a semantic graph for that snippet.
-struct flux_error_t *flux_analyze_with(struct flux_stateful_analyzer_t *, struct flux_ast_pkg_t *, struct flux_semantic_pkg_t **);
+struct flux_error_t *flux_analyze_with(struct flux_stateful_analyzer_t *, const char * src, struct flux_ast_pkg_t *, struct flux_semantic_pkg_t **);
```
I see the valgrind tests are failing in CI, my guess is they need to be updated for this change and a new tests added for the new flux_error_print function.

nathanielc commented on libflux/include/influxdata/flux.h:
```diff
@@ -27,6 +27,9 @@ void flux_free_error(struct flux_error_t *);
 const char *flux_error_str(struct flux_error_t *);
 
+// Prints the flux error to stdout
+const char *flux_error_print(struct flux_error_t *);
```
This returns a char* while the Rust implementation doesn't return anything and the Go code doesn't expect a return. Can you update this signature to match?

Do we want to print to stdout from Rust? Or should we make this method a `flux_error_format` instead that builds a string of the contents and let the caller decide to print to stdout. This would make UI errors more possible for example.

Marwes replied:
> Do we want to print to stdout from Rust? Or should we make this method a flux_error_format instead that builds a string of the contents and let the caller decide to print to stdout. This would make UI errors more possible for example.

Building a string and returning it only works for for ANSI compatible terminals so this seemed more general to me. It also auto detects if coloring is supported.

nathanielc reviewed (changes requested):
This looks great can't wait to use this. Added some comments below mostly about the changes to the C API.

Marwes commented:
> This patch is not actually changing the way errors are reported, it's just introducing the capability to use the codespan formatting, right?

No, You have to explicitly call the added formatting methods to get codespan formatting.
